README.md
# Omnipay: Pagar.Me

**Pagar.Me gateway for the Omnipay PHP payment processing library**

[![Build Status](https://img.shields.io/travis/descubraomundo/omnipay-pagarme/master.svg?style=flat-square)](https://travis-ci.org/descubraomundo/omnipay-pagarme) [![Code Climate](https://codeclimate.com/github/descubraomundo/omnipay-pagarme/badges/gpa.svg)](https://codeclimate.com/github/descubraomundo/omnipay-pagarme)  [![Test Coverage](https://codeclimate.com/github/descubraomundo/omnipay-pagarme/badges/coverage.svg)](https://codeclimate.com/github/descubraomundo/omnipay-pagarme/coverage)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/descubraomundo/omnipay-pagarme.svg?style=flat-square)](https://packagist.org/packages/descubraomundo/omnipay-pagarme)
[![Total Downloads](https://img.shields.io/packagist/dt/descubraomundo/omnipay-pagarme.svg?style=flat-square)](https://packagist.org/packages/descubraomundo/omnipay-pagarme)

[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)


[Omnipay](https://github.com/thephpleague/omnipay) is a framework agnostic, multi-gateway payment
processing library for PHP 5.3+. This package implements Pagar.Me support for Omnipay.

## Install

Via Composer

``` bash
$ composer require descubraomundo/omnipay-pagarme
```

## Basic Usage

The following gateways are provided by this package:

 * [Pagar.Me](https://pagar.me/)

For general usage instructions, please see the main [Omnipay](https://github.com/thephpleague/omnipay) repository.

### Example with Credit Card
``` php
// Create a gateway for the Pagarme Gateway
  // (routes to GatewayFactory::create)
  $gateway = Omnipay::create('Pagarme');

  // Initialise the gateway
  $gateway->initialize(array(
      'apiKey' => 'MyApiKey',
  ));

  // Create a credit card object
  // This card can be used for testing.
  $card = new CreditCard(array(
              'firstName'    => 'Example',
              'lastName'     => 'Customer',
              //'name'         => 'Example Customer',
              'birthday'     => '1988-02-28',
              'gender'       => 'M',
              'number'       => '4242424242424242',
              'expiryMonth'  => '01',
              'expiryYear'   => '2020',
              'cvv'          => '123',
              'email'        => 'customer@example.com',
              'address1'     => 'Street name, Street number, Complementary',
              'address2'     => 'Neighborhood',
              'postcode'     => '05443100',
              'phone'        => '19 3242 8855',
              'holder_document_number' => '246.375.149-23', // CPF or CNPJ
  ));

  // Do an authorize transaction on the gateway
  $transaction = $gateway->authorize(array(
      'amount'           => '10.00',
      'soft_descriptor'  => 'test', // 13 characters allowed
      'payment_method'   => 'credit_card',
      'installments'     => 5,
      'postback_url'     => 'http://application.com/api/',
      'card'             => $card,
      // 'card_hash'      => 'card_k5sT...',
      // 'card_id'        => 254786,
      'metadata'         => array(
                                'product_id' => 'ID1111',
                                'invoice_id' => 'IV2222',
                            ),
  ));
  $response = $transaction->send();
  if ($response->isSuccessful()) {
      echo "Authorize transaction was successful!\n";
      $sale_id = $response->getTransactionReference();
      $customer_id = $response->getCustomerReference();
      $card_id = $response->getCardReference();
      echo "Transaction reference = " . $sale_id . "\n";
  }
```

### Example with Boleto

``` php
// Create a gateway for the Pagarme Gateway
  // (routes to GatewayFactory::create) 
  // Create array with customer data
  $customer = array(
              'firstName'    => 'Example',
              'lastName'     => 'Customer',
              //'name'         => 'Example Customer',
              'email'        => 'customer@example.com',
              'address1'     => 'Street name, Street number, Complementary',
              'address2'     => 'Neighborhood',
              'postcode'     => '05443100',
              'phone'        => '19 3242 8855',
              'holder_document_number => '246.375.149-23', // CPF or CNPJ
  ));

  // Do an authorize transaction on the gateway
  $transaction = $gateway->authorize(array(
      'amount'           => '10.00',
      'soft_descriptor'  => 'test',
      'payment_method'   => 'boleto',
      'postback_url'     => 'http://application.com/api/',
      'customer'         => $customer,
      'metadata'         => array(
                                'product_id' => 'ID1111',
                                'invoice_id' => 'IV2222',
                            ),
  ));
  $response = $transaction->send();
  if ($response->isSuccessful()) {
      echo "Authorize Boleto transaction was successful!\n";
      $sale_id = $response->getTransactionReference();
      $boleto = $response->getBoleto();
      echo "Boleto Url = " . $boleto['boleto_url'];
      echo "Boleto Barcode = " . $boleto['boleto_barcode'];
      echo "Boleto Expiration Date = " . $boleto['boleto_expiration_date'];
      echo "Transaction reference = " . $sale_id . "\n";
  }
```


## Docs
Read the full Classes Documentation [here](http://descubraomundo.github.io/omnipay-pagarme)

## Test Mode

Pagar.Me accounts have test-mode API keys as well as live-mode API keys. 
Data created with test-mode credentials will never hit the credit card networks
and will never cost anyone money.

Unlike some gateways, there is no test mode endpoint separate to the live mode endpoint, the
Pagar.Me API endpoint is the same for test and for live. 

## Support

If you are having general issues with Omnipay, we suggest posting on
[Stack Overflow](http://stackoverflow.com/). Be sure to add the
[omnipay tag](http://stackoverflow.com/questions/tagged/omnipay) so it can be easily found.

If you want to keep up to date with release anouncements, discuss ideas for the project,
or ask more detailed questions, there is also a [mailing list](https://groups.google.com/forum/#!forum/omnipay) which
you can subscribe to.

If you believe you have found a bug, please report it using the [GitHub issue tracker](https://github.com/descubraomundo/omnipay-pagarme/issues),
or better yet, fork the library and submit a pull request.

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Change Log

## [3.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v3.0) (2019-08-26)
- Initial Relase for Omnipay v3

## [2.4.1](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v2.4.1) (2019-08-26)
- Created specific version for Omnipay v2

## [1.4.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.4.0) (2015-08-27)
**Implemented enhancements:**

- Added Request to Fetch Transaction by Id.

## [1.3.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.3.0) (2015-08-24)
**Implemented enhancements:**

- Added new Request to Calculate Installments Amount

## [1.2.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.2.0) (2015-08-20)
**Implemented enhancements:**

- Authorize, Purchase request user the Credit Card Object to set Customer Document Number.
- Create Customer Request use **holder_document_number** parameter instead of **document_number**.

## [1.1.1](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.1.1) (2015-08-18)
**Implemented enhancements:**

- The neighborhood's customer must be passed in the Address2 attribute

## [1.1.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.1.0) (2015-08-14)
**Implemented enhancements:**

- Set customer data on billet's payment method.
- Added customer key as a alternative to use the card object

**Fixed bugs:**

- Fix date format on billet expiration date

## [1.0.3](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.0.3) (2015-08-10)
**Fixed bugs:**

- Fix small bugs on Authorize Request and Response Request.

## [1.0.2](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.0.2) (2015-08-10)
**Fixed bugs:**

- Fixing PHP5.3 compatibility

## [1.0.1](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.0.1) (2015-08-07)
**Fixed bugs:**

- Fix composer configuration

## [1.0](https://github.com/descubraomundo/omnipay-pagarme/releases/tag/v1.0) (2015-08-07)
- Add tag release


src/Message/AuthorizeRequest.php
<?php
/**
 * Pagarme Authorize Request
 */

namespace Omnipay\Pagarme\Message;

/**
 * Pagarme Authorize Request
 *
 * An Authorize request is similar to a purchase request but the
 * charge issues an authorization (or pre-authorization), and no money
 * is transferred.  The transaction will need to be captured later
 * in order to effect payment. Uncaptured transactions expire in 5 days.
 *
 * Either a card object or card_id is required by default. Otherwise,
 * you must provide a card_hash, like the ones returned by Pagarme.js
 * or use the boleto's payment method.
 *
 * Pagarme gateway supports only two types of "payment_method":
 *
 * * credit_card
 * * boleto
 *
 *
 * Optionally, you can provide the customer details to use the antifraude
 * feature. These details is passed using the following attributes available
 * on credit card object:
 *
 * * firstName
 * * lastName
 * * name
 * * birthday
 * * gender
 * * address1 (must be in the format "street, street_number and neighborhood")
 * * address2 (used to specify the optional parameter "street_complementary")
 * * postcode
 * * phone (must be in the format "DDD PhoneNumber" e.g. "19 98888 5555")
 * * holder_document_number (CPF or CNPJ)
 *
 * Example:
 *
 * <code>
 *   // Create a gateway for the Pagarme Gateway
 *   // (routes to GatewayFactory::create)
 *   $gateway = Omnipay::create('Pagarme');
 *
 *   // Initialise the gateway
 *   $gateway->initialize(array(
 *       'apiKey' => 'MyApiKey',
 *   ));
 *
 *   // Create a credit card object
 *   // This card can be used for testing.
 *   $card = new CreditCard(array(
 *               'firstName'    => 'Example',
 *               'lastName'     => 'Customer',
 *               'number'       => '4242424242424242',
 *               'expiryMonth'  => '01',
 *               'expiryYear'   => '2020',
 *               'cvv'          => '123',
 *               'email'        => 'customer@example.com',
 *               'address1'     => 'Street name, Street number, Complementary',
 *               'address2'     => 'Neighborhood',
 *               'postcode'     => '05443100',
 *               'phone'        => '19 3242 8855',
 *               'holder_document_number' => '214.278.589-40',
 *   ));
 *
 *   // Do an authorize transaction on the gateway
 *   $transaction = $gateway->authorize(array(
 *       'amount'           => '10.00',
 *       'soft_descriptor'  => 'test',
 *       'payment_method'   => 'credit_card',
 *       'card'             => $card,
 *       'metadata'         => array(
 *                                 'product_id' => 'ID1111',
 *                                 'invoice_id' => 'IV2222',
 *                             ),
 *   ));
 *   $response = $transaction->send();
 *   if ($response->isSuccessful()) {
 *       echo "Authorize transaction was successful!\n";
 *       $sale_id = $response->getTransactionReference();
 *       $customer_id = $response->getCustomerReference();
 *       $card_id = $response->getCardReference();
 *       echo "Transaction reference = " . $sale_id . "\n";
 *   }
 * </code>
 *
 * @see https://docs.pagar.me/capturing-card-data/
 * @see \Omnipay\Pagarme\Gateway
 * @see \Omnipay\Pagarme\Message\CaptureRequest
 * @link https://docs.pagar.me/api/?shell#objeto-transaction
 */
class AuthorizeRequest extends AbstractRequest
{
    /**
     * Get postback URL.
     *
     * @return string
     */
    public function getPostbackUrl()
    {
        return $this->getParameter('postback_url');
    }
    
    /**
     * Set postback URL.
     *
     * @param string $value
     * @return AuthorizeRequest provides a fluent interface.
     */
    public function setPostbackUrl($value)
    {
        return $this->setParameter('postback_url', $value);
    }
    
    /**
     * Get installments.
     *
     * @return integer the number of installments
     */
    public function getInstallments()
    {
        return $this->getParameter('installments');
    }
    
    /**
     * Set Installments.
     *
     * The number must be between 1 and 12.
     * If the payment method is boleto defaults to 1.
     *
     * @param integer $value
     * @return AuthorizeRequest provides a fluent interface.
     */
    public function setInstallments($value)
    {
        return $this->setParameter('installments', (int)$value);
    }
    
    /**
     * Get soft description.
     *
     * @return string small description
     */
    public function getSoftDescriptor()
    {
        return $this->getParameter('soft_descriptor');
    }
    
    /**
     * Set soft description.
     *
     * The Pagarme gateway allow 13 characters in the soft_descriptor.
     * The provided string will be truncated if lengh > 13.
     *
     * @param string $value
     * @return AuthorizeRequest provides a fluent interface.
     */
    public function setSoftDescriptor($value)
    {
        return $this->setParameter('soft_descriptor', substr($value, 0, 13));
    }
    
    /**
     * Get the boleto expiration date
     *
     * @return string boleto expiration date
     */
    public function getBoletoExpirationDate($format = 'Y-m-d\TH:i:s')
    {
        $value = $this->getParameter('boleto_expiration_date');
        
        return $value ? $value->format($format) : null;
    }
    
    /**
     * Set the boleto expiration date
     *
     * @param string $value defaults to atual date + 7 days
     * @return AuthorizeRequest provides a fluent interface
     */
    public function setBoletoExpirationDate($value)
    {
        if ($value) {
            $value = new \DateTime($value, new \DateTimeZone('UTC'));
            $value = new \DateTime($value->format('Y-m-d\T03:00:00'), new \DateTimeZone('UTC'));
        } else {
            $value = null;
        }
        
        return $this->setParameter('boleto_expiration_date', $value);
    }
    
    public function getData()
    {
        $this->validate('amount');
        
        $data = array();
        
        $data['amount'] = $this->getAmountInteger();
        $data['payment_method'] = $this->getPaymentMethod();
        $data['postback_url'] = $this->getPostbackUrl();
        $data['installments'] = $this->getInstallments();
        $data['soft_descriptor'] = $this->getSoftDescriptor();
        $data['metadata'] = $this->getMetadata();
        if ($this->getPaymentMethod() && ($this->getPaymentMethod() == 'boleto')) {
            if ($this->getBoletoExpirationDate()) {
                $data['boleto_expiration_date'] = $this->getBoletoExpirationDate();
            }
            $data['payment_method'] = $this->getPaymentMethod();
            if ($this->getCard()) {
                $data = array_merge($data, $this->getCustomerData());
            } elseif ($this->getCustomer()) {
                $this->setCard($this->getCustomer());
                $data = array_merge($data, $this->getCustomerData());
            }
        } else {
            if ($this->getCard()) {
                $data = array_merge($data, $this->getCardData(), $this->getCustomerData());
            } elseif ($this->getCardHash()) {
                $data['card_hash'] = $this->getCardHash();
            } elseif ($this->getCardReference()) {
                $data['card_id'] = $this->getCardReference();
            } else {
                $this->validate('card');
            }
        }
        $data['capture'] = 'false';
        
        return $data;
    }
    
    public function getEndpoint()
    {
        return $this->endpoint.'transactions';
    }
}
